README.md
# d-lev-librarian
 Source mirror for D-Lev librarian software, hosted at https://d-lev.com/support.html.

# Instructions for use
 Run `./build_all.sh`, and refer to original documentation in `.txt` files. 
 This is a user-facing piece of software. Among its other functions, it transfers the compiled `.spi` payload generated by [d-lev-software](https://github.com/d-lec/d-lev-software) to the FPGA, through a standard UART<->USB bridge.  
 Please refer to the [D-Lev manual](http://d-lev.com/support/D-Lev_Manual_2023-06-20.pdf) for instructions.  

# Repo conventions
"Original text" for this repo can be found at `http://d-lev.com/support/d-lib_[RELEASE_DATE-YYYY-MM-DD].zip`. When mirroring an update from the official D-Lev website, please make a "git tag" corresponding to the release date, and upload the original zip file from the website to the Releases page.  


CHANGE_LOG.txt
///////////
// BUILD //
///////////
- Linux:   GOOS=linux GOARCH=amd64 go build -o d-lin
- Windows: GOOS=windows GOARCH=amd64 go build -o d-win.exe
- Mac:     GOOS=darwin GOARCH=amd64 go build -o d-mac
- Mac M1:  GOOS=darwin GOARCH=arm64 go build -o d-mm1
- ARM64:   GOOS=linux GOARCH=arm64 go build -o d-arm
- ARM32:   GOOS=linux GOARCH=arm go build -o d-a32

- build_all.sh : all of the above
- build.sh     : just Linux (for my dev.)

////////////////
// CHANGE LOG //
////////////////
2023-06-18 : v8
- Command update: 
  - Now the default at startup.
  - _ALL_ dir now relative to exe location.
  - Option flag -d sets temp work dir.
- Command ver reports librarian & SW ver, does CRC & old version checks.
- A bunch of older SW versions now in d-lib_pkg.go.
- Command help now has -v verbose flag.

2023-06-17 : v8
- Serial port idenfier is now system name rather than number from list.
- Serial port no longer managed by main(), port open reads config file.
- More verbose config file interaction.
- SW_NEW, SW_OLD, DATE now in d-lib_pkg.go.
- file_write() => file_write_str() with trim.
- get_file_str() => file_read_str(), file_read_chk().
- Funneled all file reads through file_read_str().
- Funneled all file writes through file_write_str().
- spi_wr() mode unused, removed.

2023-06-16 : v8
- Updated go to v1.20.4 and go.bug.st/serial to v1.5.0 (mac serial issues).
- Rearranged PITCH UI knobs.
- sp_wr_rd() issues ResetInputBuffer() & ResetOutputBuffer().
- Command join safer, uses get_file_str().
- get_file_str() safer, trims spaces at ends.

2023-06-10 : v7 - RELEASED as 2023-05-24 - 
- fmt.Scanln() stops consuming at first space, leaving stuff in the buffer.
- Replaced with user_word() using bufio.NewScanner(os.Stdin) & .Scan().

2023-06-08 : v7
- Still more work on the update menu.

2023-06-06 : v7
- Command menu => update.

2023-06-05 : v7
- Command update => menu.
- New file "lib_time.go".

2023-06-03 : v7
- update command walks user thru the steps.
- New file "lib_cmd.go" contains most main() code.

2023-06-01 : v7
- Removed stob command, expanded match command.
- batch command -r (=> Rob S.) tweaked w/ Rob here.
- batch command -m (=> mono) fixed.
- batch command -s (=> stereo) removed.

2023-05-28 : v7
- batch command for DLP batch processing.
- Added -y flag to all file writing commands to bypass <y|n> prompt (for scripting, h/t Craig).
- Added directory check & creation for commands ktof, stof, dump.
- Funneled most file writes through file_write().

2023-05-17 : v7
- diff command can compare two slots, knobs to slot
- stob command flag: -i => -g (guess)

2023-05-15 : v7
- match command flag: -i => -g (guess)
- knob command flag: -i => -o (offset)

2023-05-11b : v7
- update command now fixes preview knobs.

2023-05-11 : v7
- Edited to reflect pitch preview changes:
- pp_p5 : "treb" type 0xc5 => "tone" type 0xc5
- pp_p6 : "bass" type 0xc5 => "vmod" type 0xca

2023-05-10 : v7
- slots => "match": match slots or dir to dir of *.dlp files.
- knob command flag -k combines page_name:knob_number; -p flag removed.
- slots command: removed use of file map when -inf=false.

2023-05-09b : v7
- New knob command to read, +/-, set value of knob.
- morph seed flag: -seed => -i
- slots infer flag: -inf => -i
- stob infer flag: -inf => -i
- stob header flag: -hdr => -h

2023-03-13 : v7
- Build now targets ARM32 & 64 bit Linux OS (32 bit tested on RPI4).

2023-01-31 : v6 - RELEASED -
- Command morph now has -mo, -mn, -me -mf, -mr scalings, -seed assignment.
- Command slots column width now adapts to file name lengths.

2023-01-29 : v6
- Fleshing out morph command.

2023-01-28 : v6
- New join command.
- Update command now skips blank files.
- Bug: ints_to_hexs was incorrectly trimming last '\n' => bad DLP concat.

2023-01-25 : v6
- Changes to reflect SW update.

2023-01-17 : v6
- Some refactoring in preset code.
- Slots command: restored use of file map when -inf=false (faster).

2023-01-15 : v6
- Changes to reflect type shuffling & changes.

2023-01-12 : v6
- Loop command: more robust, removed timer.

2023-01-10 : v6
- Ports command: slight change to text output.
- Help text edited a bit (removed flags "=").

2023-01-09 : v6 - RELEASED & WITHDRAWN -
- New loop command: test raw dongle.
- Split command: now deposits files in source directory.
- Update command: now reports de-fluff count.

2023-01-04 : v6
- Slots command: new -inf flag to return best guess if no exact match.
- Some cleaning & generalization of file subs.

2023-01-01 : v5 - RELEASED -
- Update command: pitch correction & stereo normalizations.

2022-12-30 : v5
- Update command: NOISE:vmod and RESON:mode changes.

2022-12-28 : v5
- Diff command: now compares a file to knobs / slot / file.

2022-12-27 : v5
- Update command: convert all presets to latest arch (undocumented).
- Diff command: only displays screens with differences.

2022-12-24 : v5
- New "diff" command to compare two preset files.
- View command now shows blank pre knob values for pro, & vice versa.

2022-12-23 : v5
- Commands stob, pump, dump, split: -f <file> flag added for consistency.
- Command ports: -p <port> flag added for safety / clarity.
- Command slots: -d <directory> flag added for clarity.
- Command stob: -hdr flag to enable file header comments.
- Sprinkled around blank file name checks for safety.
- More array bounds checking to hopefully prevent panics.
- All -s <slot> values now strings to defeat default behavior.
- Port setting logic more robust.

2022-12-19 : v5
- Bank file header comments every 10 slots.

2022-12-17 : v5
- Changes to reflect [0:249] presets & [0:5] profiles.
- New "reset" command to reset the D-Lev processor.
- New "split" command to parse container files to their sub-components.
- Commands split, pump, dump use the file extension to know what to do.
- Command stob -f flag removed (redundant).

2022-12-14 : v4
- Added reset command, issued automatically after spi | eeprom pump.
- Changes to reflect type shuffling & changes.

d-lib_pkg.go
package main

/*
 * d-lev constants & helper functions
*/

import (
	"strings"
	"strconv"
	"math/bits"
	"fmt"
	"os"
)

type ver_tbl_t struct {
	lib string
	sw string
	date string
}

// librarian & software versions, dates
// current @ [0]
var ver_tbl = []ver_tbl_t {  
	{"9",		"6be9394f",	"2023-07-26"}, // 0
	{"8",		"7bbb846b",	"2023-06-20"}, // 1
	{"7",		"73c6c3d7",	"2023-05-24"}, // 2
	{"6",		"27c263bf",	"2023-01-31"}, // 3
	{"5",		"2d58f653",	"2023-01-01"}, // 4
	{"2",		"add46826",	"2022-10-06"}, // 5
	{"OV129",	"7bc1bd55",	"2022-07-05"},
	{"OV128",	"93152c8b",	"2022-05-10"},
	{"OV127",	"d202d35",	"2022-05-04"},
	{"OV126",	"af3f63c4",	"2022-04-30"},
	{"OV125",	"67517a97",	"2022-04-17"},
	{"OV124",	"5ba55477",	"2022-03-17"},
	{"OV121",	"7b6a0484",	"2022-01-01"},
	{"OV120",	"84f7f31c",	"2021-12-18"},
	{"OV119",	"52fe7d",	"2021-12-04"},
	{"OV115",	"240b1e68",	"2021-10-31"},
}

const (
	SLOTS = 256											// pre + pro slots
	SLOT_BYTES = 256									// bytes per slot
	PRO_SLOTS = 6										// profile[0:5]
	PRE_SLOTS = SLOTS - PRO_SLOTS						// preset[0:249]
	//
	EE_RW_BYTES = 4										// eeprom bytes per read / write cycle
	EE_PG_BYTES = 256									// eeprom bytes per page
	//
	EE_PRE_ADDR = 0x0									// eeprom pre start addr
	EE_PRE_END = EE_PRE_ADDR + (PRE_SLOTS * SLOT_BYTES)	// eeprom pre end addr
	//
	EE_PRO_ADDR = EE_PRE_END							// eeprom pro start addr
	EE_PRO_END = EE_PRO_ADDR + (PRO_SLOTS * SLOT_BYTES)	// eeprom pro end addr
	//
	EE_SPI_ADDR = EE_PRO_END							// eeprom code start addr
	EE_SPI_SZ = 0x4000									// eeprom code size : 16kB code space
	EE_SPI_END = EE_SPI_ADDR + EE_SPI_SZ				// eeprom code end addr
	//
	EE_START = EE_PRE_ADDR								// eeprom start addr
	EE_END = EE_SPI_END									// eeprom end addr
	EE_WR_MS = 6										// eeprom write wait time (ms)
	//
	PAGES_COLS = 4										// pages print columns
	PAGES_ROWS = 5										// pages print rows
	PAGES = PAGES_COLS * PAGES_ROWS						// pages
	//
	KNOBS_COLS = 2										// knob columns
	KNOBS_ROWS = 4										// knob rows
	KNOBS = KNOBS_COLS * KNOBS_ROWS						// knobs
	KNOBS_TOTAL = KNOBS * PAGES							// total knobs
	PAGE_SEL_KNOB = 7									// page selector knob
	//
	RX_BUF_BYTES = 512									// serial port rx buffer size
	CHARS_PER_DOT = 4096								// chars for each activity dot printed
	CFG_FILE = "d-lib.cfg"								// config file name
	WORK_DIR = "_WORK_"									// work scratch dir
	PRESETS_DIR = "_ALL_"								// presets dir
	CRC = "debb20e3"									// good CRC
)

// given sw_ver, return date
func sw_date_lookup(sw_ver string) (string) {
	for _, entry := range ver_tbl {
		if sw_ver == entry.sw { return entry.date }
	}
	return "unknown"
}

// convert string of multi-byte hex values to slice of ints
// hex string values on separate lines
func hexs_to_ints(hex_str string, bytes int) ([]int) {
	var ints []int
	str_split := (strings.Split(strings.TrimSpace(hex_str), "\n"))
	for _, str := range str_split {
		sh_reg, err := strconv.ParseUint(str, 16, 32); err_chk(err)
		for b:=0; b<bytes; b++ { 
			sh_byte := int(uint8(sh_reg))
			ints = append(ints, sh_byte)
			sh_reg >>= 8
		}
	}
	return ints
}

// convert slice of ints to string of multi-byte hex values
// hex string values on separate lines
func ints_to_hexs(ints []int, bytes int) (string) {
	var hex_str string
	for i:=0; i<len(ints); i+=bytes {
		var line_int int64
		for b:=0; b<bytes; b++ { 
			line_int += int64(uint8(ints[i+b])) << (b * 8)
		}
		hex_str += strconv.FormatInt(line_int, 16) + "\n"
	}
	return hex_str
}

// check for hexness
func str_is_hex(str string) bool {
	if len(str) == 0 { return false }
	for _, ch := range str {
		if !((ch >= '0' && ch <= '9') || (ch >= 'a' && ch <= 'f') || (ch >= 'A' && ch <= 'F')) { return false }
	}
	return true
}

// return index of string in slice, else -1
func str_exists(strs []string, str string) (int) {
	for idx, entry := range strs { if str == entry { return idx } }
	return -1
}

// return crc of 32 bit input
func crc_32(sh_reg uint32) (uint32) {
	poly := uint32(0x6db88320)
	for i:=0; i<32; i++ { 
		sh_reg = bits.RotateLeft32(sh_reg, -1)  // >>r 1
		if sh_reg & 0x80000000 != 0 { sh_reg ^= poly }  // xor w/ poly if MSb is set
	}
	return sh_reg
}

// print quit message and exit program
func quit_exit() {
	fmt.Println("> -QUIT- exiting program...")
	os.Exit(0) 
}

// print error message and exit program
func error_exit(error_str string) {
	fmt.Println("> -ERROR-", error_str, "!")
	os.Exit(0) 
}

// check for error, exit program if true
func err_chk(err error) {
	if err != nil { error_exit(err.Error()) }
}
